### PR TITLE
[FLINK-34555][table] Migrate JoinConditionTypeCoerceRule to java.

### PR DESCRIPTION
**Pull request by liuyongvs**
## What is the purpose of the change

The PR migrates JoinConditionTypeCoerceRule to java
it doesn't touch JoinConditionTypeCoerceRuleTest to be sure that java version continues passing it

## Verifying this change

This change is already covered by existing tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by snuyanzin** [comment]
@liuyongvs the ci still fails could you please have a look here?

**Comment by liuyongvs** [comment]
> @liuyongvs the ci still fails could you please have a look here?

@snuyanzin The error occurs because the word "implicit" was mistakenly written as "Implicit", which leads to a compilation error in testLogicalPlanWithImplicitTypeCast,  have fixed it

**Comment by liuyongvs** [comment]
hi @snuyanzin will you help review it?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/JoinConditionTypeCoerceRule.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        RelDataTypeFactory typeFactory = builder.getTypeFactory();
+
+        List<RexNode> joinFilters = RelOptUtil.conjunctions(join.getCondition());
+        List<RexNode> newJoinFilters =
+                joinFilters.stream()
+                        .map(
+                                filter -> {
+                                    if (filter instanceof RexCall) {
+                                        RexCall c = (RexCall) filter;
+                                        if (c.getKind().equals(SqlKind.EQUALS)) {
```
Comment (snuyanzin) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/JoinConditionTypeCoerceRule.java:87]:
```suggestion
                                        if (c.getKind() == SqlKind.EQUALS) {
```
it's better to follow same approach in code
e.g. below it is compared with `==` which is preferred for  enums

**Comment by liuyongvs** [comment]
hi @snuyanzin ci passed

**Comment by liuyongvs** [comment]
hi @snuyanzin , will you have time to review it again?

**Review by snuyanzin — APPROVED** [review]

**Pull request merged**